Source/Core/Common/BlockingLoop.h

```diff
@@ -132,64 +133,46 @@ class BlockingLoop
 
     while (!m_shutdown.IsSet())
     {
-      payload();
+      if (m_running_state.load() == RunningState::NeedExecution)
+      {
+        m_running_state.store(RunningState::Executing);
+
+        payload();
 
-      switch (m_running_state.load())
+        // If we're still in the Running state, then Wakeup wasn't called within the last execution
+        // of the payload. This means we should be ready now. Else we're in the Done state now, so
+        // wakeup the waiting threads right now.
+        RunningState expected_running{RunningState::Executing};
+        if (m_running_state.compare_exchange_strong(expected_running, RunningState::Done))
```

Comment:
If we accept the executing state in line 136, we could use the weak exchange variant here as well.

Reply:
Can you explain that more? If the loop is in RunningState::Executing and spuriously fails to exchange that for RunningState::Done, then goes back to the toplevel, it'll go into the busy loop/done state without ever signifying to the user that's it's done... I think.